src/inmanta/references.py

```diff
@@ -0,0 +1,450 @@
+            del type(self)._reference_classes[self.name]
+
+        type(self)._reference_classes[self.name] = cls
+        cls.type = self.name
+        return cls
+
+    @classmethod
+    def get_class(cls, name: str) -> type[Reference[RefValue]]:
+        """Get the class of registered with the given name"""
+        if name not in cls._reference_classes:
+            raise TypeError(f"There is no reference class registered with name {name}")
+
+        return cls._reference_classes[name]
+
+    @classmethod
+    def reset(cls) -> None:
+        """Reset the registered reference classes"""
+        cls._reference_classes = {}
+
+    @classmethod
+    def get_references(cls) -> typing.Iterator[tuple[str, type[Reference[RefValue]]]]:
+        """Return an iterator with all items registered."""
+        return (item for item in cls._reference_classes.items())
+
+
+class mutator:
+    """This decorator register a mutator under a specific name"""
+
+    _mutator_classes: typing.ClassVar[dict[str, type[Mutator]]] = {}
+
+    def __init__(self, name: str) -> None:
+        """
+        :param name: This name is used to indicate the type of the mutator
+        """
+        self.name = name
+
+    def __call__[T: Mutator](self, cls: type[T]) -> type[T]:
+        """Register a new mutator. If we already have it explicitly delete it (reload)"""
+        if self.name in mutator._mutator_classes:
+            del mutator._mutator_classes[self.name]
+
+        mutator._mutator_classes[self.name] = cls
+        cls.type = self.name
+        return cls
+
+    @classmethod
+    def get_class(cls, name: str) -> type[Mutator]:
+        """Get the class of registered with the given name"""
+        if name not in cls._mutator_classes:
+            raise TypeError(f"There is no mutator class registered with name {name}")
+
+        return cls._mutator_classes[name]
+
+    @classmethod
+    def reset(cls) -> None:
+        """Reset the registered mutator classes"""
+        cls._mutator_classes = {}
+
+    @classmethod
+    def get_mutators(cls) -> typing.Iterator[tuple[str, type[Mutator]]]:
+        """Return an iterator with all items registered."""
+        return (item for item in cls._mutator_classes.items())
+
+
+@reference(name="core::AttributeReference")
+class AttributeReference[T: PrimitiveTypes](Reference[T]):
+    """A reference that points to a value in an attribute of a dataclass"""
+
+    def __init__(
+        self,
+        reference: DataclassProtocol | DataclassReference[DataclassProtocol],
+        attribute_name: str,
+        attribute_type: typing.Type[T],
+    ) -> None:
+        super().__init__(reference=reference, attribute_name=attribute_name, attribute_type=attribute_type)
+        self.attribute_name = attribute_name
+        self.attribute_type = attribute_type
+        self.reference = reference
+
+    def resolve(self) -> T:
+        return typing.cast(T, getattr(self.reference, self.attribute_name))
```

Comment:
Since we transparently skip what would otherwise be model validation (e.g. when assigned to a resource's attribute, should we add some validation here?